### PR TITLE
Add: support uneven DSpark decode owner occupancy

### PR DESCRIPTION
**Pull request by sjduan**
- Pass per-owner active-prefix counts through decode forward and
  skip attention and the LM head only for an entirely empty TP group.
- Keep all EP ranks in MoE communication, including zero-token gate
  owners, and mask inactive embeddings, outputs and cache metadata.
- Zero every empty-group logits row while retaining sampled_ids=-1.
- Handle fully invalid CSA sparse windows and add fixture controls
  for uneven occupancy and all-invalid attention rows.

### CODE REVIEW

**Comment by sjduan** [comment]
Validation status: this PR is intentionally **Draft**, not device-accepted.

- Base: `645166036bf2b3f84a3df7053c20a40c3df403bf`; its only change after the source rebase at `774b46c` is merge-workflow documentation.
- Each isolated PR passed repository pre-commit checks, Python syntax, 44 checked cross-module call signatures, and DP4/TP4/EP16 module imports in the `dsj-pypto-dev` environment. The RoPE PR also passed its three contract tests.
- The rebased **combined** lib changes plus the matching local serving patch passed 59 host tests (56 serving regressions + 3 lib RoPE contracts). This does not establish standalone device correctness of an isolated PR.
- Earlier device evidence belongs to the predecessor combined working tree: component regressions and TP2/EP2 full43 runtime witnesses, including zero/partial owners. It is historical evidence, not a fresh result on this base.
- New device regressions are pending because the test queue is in maintenance mode. No queue locks were bypassed, no NPU case was run for this rebase, and no heap/toolchain workaround was added. Before promotion, rerun affected component/TP paths and the combined persistent-serving gate with unchanged numerical thresholds.

Remaining issues stay open in #1133; this patch series does not close that tracker.

Serving coordination: `num_tokens_per_owner` is one INT32 active-prefix count per EP rank. Physical owner shapes stay fixed; each count is an S-multiple. Every rank in a nonempty TP group must join attention collectives, and all EP ranks still join MoE. Serving must mask inactive write slots/lens consistently. This is not DP16 support; DP4/TP4/EP16 remains the target. Combine with #1134, #1135, #1136 and #1137 before the persistent-serving gate. Shared `decode_fwd.py` hunks must retain RoPE profile routing and cache-line packing when merged.

File: models/deepseek_v4_flash_dspark/decode_fwd.py [hunk 1]
```diff
@@ -1046,34 +1151,45 @@ def decode_fwd(
...
+                pl.const(LM_HEAD_COMM_EPOCH, pl.INT32),
+                final_norm_tid,
+            )
+            greedy_sample(logits, sampled_ids)
+            mask_inactive_sample_rows(logit_row_indices, sampled_ids)
+        else:
+            for row in pl.spmd(MAX_LOGIT_ROWS, name_hint="decode_fwd_inactive_sample_rows"):
+                sampled_ids[row : row + 1, :] = pl.full(
+                    [1, SAMPLED_IDS_PAD], dtype=pl.INT32, value=-1,
+                )
```
[the remark this answers is not in the document]
Comment (sjduan) [reply models/deepseek_v4_flash_dspark/decode_fwd.py:1202]: Fixed in a902bc66. The empty-group branch now zero-fills every logits row using the existing MAX_LOGIT_ROWS loop, while preserving sampled_ids=-1. The clear uses 256-column tiles (with a vocabulary tail) instead of a full-vocabulary on-core tensor; the nonempty path and public ABI are unchanged.

Validation: the exact production loop, AST-checked in a private regression, passed on a2a3 with poisoned InOut buffers and zero tolerance: logits [512,129280] became all zero and sampled_ids [512,8] became all -1. A tail-width case [4,129408] also passed. A TP2/EP2 all-empty forward smoke (--start-pos 120 --num-tokens-per-owner 0) passed compile/runtime; its CLI has no golden, so this is not a full-model numerical claim. Tests used PTOAS 0.57. Three host contracts, syntax/diff checks, and all pre-commit hooks passed. No probe/debug file was added to the PR.
